From 3125a8bc1754e1a24f2907a42fb46c6eb4f3908f Mon Sep 17 00:00:00 2001
From: Bai Ping <ping.bai@nxp.com>
Date: Fri, 10 Mar 2017 14:54:01 +0800
Subject: [PATCH 1/2] MLK-14409-01 ARM: imx: Add speed grading fuse check for
 900MHz on i.mx6ull

According to the latest datasheet(Rev.1,02/2017), when the internal LDO
is enabled, the ARM core can run at 900MHz. We need to check the
speed grading fuse to determine the max ARM core frequency.

Signed-off-by: Bai Ping <ping.bai@nxp.com>
---
 arch/arm/mach-imx/mach-imx6ul.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/arch/arm/mach-imx/mach-imx6ul.c b/arch/arm/mach-imx/mach-imx6ul.c
index f20a9fa..84d202b 100644
--- a/arch/arm/mach-imx/mach-imx6ul.c
+++ b/arch/arm/mach-imx/mach-imx6ul.c
@@ -59,6 +59,7 @@ static void __init imx6ul_enet_phy_init(void)
 #define OCOTP_CFG3			0x440
 #define OCOTP_CFG3_SPEED_SHIFT		16
 #define OCOTP_CFG3_SPEED_696MHZ		0x2
+#define OCOTP_CFG3_SPEED_900MHZ		0x3
 
 static void __init imx6ul_opp_check_speed_grading(struct device *cpu_dev)
 {
@@ -87,7 +88,7 @@ static void __init imx6ul_opp_check_speed_grading(struct device *cpu_dev)
 	 * 2b'00: Reserved;
 	 * 2b'01: 528000000Hz;
 	 * 2b'10: 700000000Hz(i.MX6UL), 800000000Hz(i.MX6ULL);
-	 * 2b'11: Reserved;
+	 * 2b'11: 900000000Hz(i.MX6ULL);
 	 * We need to set the max speed of ARM according to fuse map.
 	 */
 	val = readl_relaxed(base + OCOTP_CFG3);
@@ -105,6 +106,11 @@ static void __init imx6ul_opp_check_speed_grading(struct device *cpu_dev)
 			if (dev_pm_opp_disable(cpu_dev, 792000000))
 				pr_warn("Failed to disable 792MHz OPP\n");
 		}
+
+		if (val != OCOTP_CFG3_SPEED_900MHZ) {
+			if(dev_pm_opp_disable(cpu_dev, 900000000))
+				pr_warn("Failed to disable 900MHz OPP\n");
+		}
 	}
 	iounmap(base);
 
-- 
1.9.1

